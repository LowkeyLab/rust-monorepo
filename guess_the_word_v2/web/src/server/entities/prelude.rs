//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.12

pub use super::game_players::Entity as GamePlayers;
pub use super::games::Entity as Games;
pub use super::players::Entity as Players;
pub use super::rounds::Entity as Rounds;
