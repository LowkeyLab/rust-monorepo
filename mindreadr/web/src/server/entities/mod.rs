//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.12

pub mod prelude;

pub mod game_players;
pub mod games;
pub mod round_guesses;
pub mod rounds;
pub mod sea_orm_active_enums;
