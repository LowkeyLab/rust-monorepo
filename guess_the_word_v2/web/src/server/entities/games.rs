//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.12

use sea_orm::entity::prelude::*;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "games")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    pub player_id: String,
    pub word: String,
    pub guesses: Json,
    pub status: String,
    pub score: Option<i32>,
    pub max_attempts: i32,
    pub created_at: DateTime,
    pub updated_at: DateTime,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}
