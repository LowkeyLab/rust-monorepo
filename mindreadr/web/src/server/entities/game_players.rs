//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.12

use sea_orm::entity::prelude::*;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "game_players")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub game_id: i32,
    #[sea_orm(primary_key, auto_increment = false)]
    pub player_id: i32,
    pub joined_at: DateTime,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::games::Entity",
        from = "Column::GameId",
        to = "super::games::Column::Id",
        on_update = "NoAction",
        on_delete = "Cascade"
    )]
    Games,
    #[sea_orm(
        belongs_to = "super::players::Entity",
        from = "Column::PlayerId",
        to = "super::players::Column::Id",
        on_update = "NoAction",
        on_delete = "Cascade"
    )]
    Players,
}

impl Related<super::games::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Games.def()
    }
}

impl Related<super::players::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Players.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
