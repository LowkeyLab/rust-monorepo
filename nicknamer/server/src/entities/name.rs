//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.12

use sea_orm::entity::prelude::*;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "name")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    #[sea_orm(unique)]
    pub discord_id: i64,
    pub name: String,
    pub server_id: String,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}
