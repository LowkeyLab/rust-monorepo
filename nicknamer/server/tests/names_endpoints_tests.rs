use axum::body::Body;
use axum::http::{Method, Request, StatusCode};
use insta::assert_yaml_snapshot;
use nicknamer_server::entities::name;
use nicknamer_server::name::{NameState, create_name_router};
use sea_orm::{ActiveModelTrait, DatabaseConnection, Set};
use serde::Serialize;
use std::collections::BTreeMap;
use std::sync::Arc;
use testcontainers_modules::{postgres, testcontainers};
use tower::ServiceExt;

mod common;

/// HTTP response snapshot for testing endpoints.
#[derive(Debug, Serialize)]
struct HttpResponseSnapshot {
    test_context: String,
    status: u16,
    content_type: Option<String>,
    headers: std::collections::BTreeMap<String, String>,
    html_body: Vec<String>,
}

impl HttpResponseSnapshot {
    /// Create a new HTTP response snapshot.
    fn new(
        body_text: &str,
        status: StatusCode,
        content_type: Option<&str>,
        headers: &axum::http::HeaderMap,
        test_context: &str,
    ) -> Self {
        Self {
            test_context: test_context.to_string(),
            status: status.as_u16(),
            content_type: content_type.map(|s| s.to_string()),
            headers: filter_non_time_sensitive_headers(headers),
            html_body: normalize_html_for_snapshot(body_text),
        }
    }
}

/// Normalize HTML content for consistent snapshots by removing dynamic values.
fn normalize_html_for_snapshot(html: &str) -> Vec<String> {
    // Split HTML by newlines and convert to Vec<String>
    // In the future, we could add more sophisticated normalization
    html.lines().map(|line| line.to_string()).collect()
}

/// Filter out time-sensitive headers from response headers for snapshot testing.
fn filter_non_time_sensitive_headers(headers: &axum::http::HeaderMap) -> BTreeMap<String, String> {
    let time_sensitive_headers = [
        "date",
        "expires",
        "last-modified",
        "etag",
        "server",
        "x-request-id",
        "x-trace-id",
        "set-cookie",
    ];

    headers
        .iter()
        .filter_map(|(name, value)| {
            let name_str = name.as_str().to_lowercase();
            if time_sensitive_headers.contains(&name_str.as_str()) {
                None
            } else {
                value.to_str().ok().map(|v| (name_str, v.to_string()))
            }
        })
        .collect()
}

/// Test context for endpoint tests.
pub struct TestContext {
    #[allow(dead_code)] // container is kept to ensure it's not dropped
    pub container: testcontainers::ContainerAsync<postgres::Postgres>,
    pub db: DatabaseConnection,
}

/// Setup function for endpoint tests using PostgreSQL container.
async fn setup() -> anyhow::Result<TestContext> {
    // Allow multiple calls to init for tests.
    let _ = tracing_subscriber::fmt().try_init();
    let container = common::setup_container().await?;
    let db = common::setup_db(&container).await?;
    Ok(TestContext { db, container })
}

/// Test helper to create test names in the database.
async fn create_test_names(db: &DatabaseConnection) {
    let name1 = name::ActiveModel {
        discord_id: Set(123456789),
        name: Set("TestUser1".to_string()),
        ..Default::default()
    };

    let name2 = name::ActiveModel {
        discord_id: Set(987654321),
        name: Set("TestUser2".to_string()),
        ..Default::default()
    };

    let _result1 = name1.insert(db).await.unwrap();
    let _result2 = name2.insert(db).await.unwrap();
}

/// Test helper to create a single test name and return its ID.
async fn create_single_test_name(db: &DatabaseConnection) -> i32 {
    let name = name::ActiveModel {
        discord_id: Set(555444333),
        name: Set("DeleteTestUser".to_string()),
        ..Default::default()
    };

    let result = name.insert(db).await.unwrap();
    result.id
}

#[tokio::test]
async fn can_display_names_table_when_names_exist() {
    let state = setup().await.expect("Failed to setup test context");
    create_test_names(&state.db).await;

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state);

    let request = Request::builder()
        .uri("/names")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "names_table_with_existing_names",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_display_empty_names_table_when_no_names_exist() {
    let state = setup().await.expect("Failed to setup test context");

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state);

    let request = Request::builder()
        .uri("/names")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "empty_names_table",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn names_endpoint_returns_correct_content_type() {
    let state = setup().await.expect("Failed to setup test context");

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state);

    let request = Request::builder()
        .uri("/names")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);
    assert_eq!(
        response.headers().get("content-type").unwrap(),
        "text/html; charset=utf-8"
    );

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "names_endpoint_content_type_check",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_create_name_successfully() {
    let state = setup().await.expect("Failed to setup test context");

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state.clone());

    let form_data = "discord_id=555666777&name=NewTestUser";
    let request = Request::builder()
        .method(Method::POST)
        .uri("/names")
        .header("content-type", "application/x-www-form-urlencoded")
        .body(Body::from(form_data))
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "create_name_successfully",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_create_multiple_names_and_update_count() {
    let state = setup().await.expect("Failed to setup test context");
    create_test_names(&state.db).await;

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state.clone());

    let form_data = "discord_id=111222333&name=ThirdUser";
    let request = Request::builder()
        .method(Method::POST)
        .uri("/names")
        .header("content-type", "application/x-www-form-urlencoded")
        .body(Body::from(form_data))
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "create_multiple_names_update_count",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_handle_form_with_special_characters_in_name() {
    let state = setup().await.expect("Failed to setup test context");

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state.clone());

    let form_data = "discord_id=888999000&name=User%20With%20Spaces%21";
    let request = Request::builder()
        .method(Method::POST)
        .uri("/names")
        .header("content-type", "application/x-www-form-urlencoded")
        .body(Body::from(form_data))
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "form_with_special_characters",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_serve_add_name_form() {
    let state = setup().await.expect("Failed to setup test context");

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state);

    let request = Request::builder()
        .uri("/names/form")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);
    assert_eq!(
        response.headers().get("content-type").unwrap(),
        "text/html; charset=utf-8"
    );

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "add_name_form",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn post_endpoint_returns_table_fragment_not_full_page() {
    let state = setup().await.expect("Failed to setup test context");

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state.clone());

    let form_data = "discord_id=777888999&name=FragmentTestUser";
    let request = Request::builder()
        .method(Method::POST)
        .uri("/names")
        .header("content-type", "application/x-www-form-urlencoded")
        .body(Body::from(form_data))
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "table_fragment_not_full_page",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn cannot_create_name_with_duplicate_discord_id() {
    let state = setup().await.expect("Failed to setup test context");

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state.clone());

    // First, create a name with a specific Discord ID
    let form_data = "discord_id=123456789&name=FirstUser";
    let request = Request::builder()
        .method(Method::POST)
        .uri("/names")
        .header("content-type", "application/x-www-form-urlencoded")
        .body(Body::from(form_data))
        .unwrap();

    let response = app.oneshot(request).await.unwrap();
    assert_eq!(response.status(), StatusCode::OK);

    // Now try to create another name with the same Discord ID
    let duplicate_form_data = "discord_id=123456789&name=SecondUser";
    let app2 = create_name_router(name_state.clone());
    let duplicate_request = Request::builder()
        .method(Method::POST)
        .uri("/names")
        .header("content-type", "application/x-www-form-urlencoded")
        .body(Body::from(duplicate_form_data))
        .unwrap();

    let duplicate_response = app2.oneshot(duplicate_request).await.unwrap();

    let headers = duplicate_response.headers().clone();
    let body = axum::body::to_bytes(duplicate_response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    // Verify the error message is user-friendly
    assert!(body_text.contains("A name entry already exists for this Discord ID"));

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::UNPROCESSABLE_ENTITY,
        Some("text/html; charset=utf-8"),
        &headers,
        "duplicate_discord_id_error",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_delete_name_successfully() {
    let state = setup().await.expect("Failed to setup test context");
    let name_id = create_single_test_name(&state.db).await;

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state);

    let request = Request::builder()
        .method(Method::DELETE)
        .uri(format!("/names/{}", name_id))
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    // Should return the updated names table (empty in this case)
    assert!(body_text.contains("No names found in the database"));

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "delete_name_successfully",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_delete_name_and_update_table_count() {
    let state = setup().await.expect("Failed to setup test context");

    // Create multiple names
    create_test_names(&state.db).await;
    let delete_name_id = create_single_test_name(&state.db).await;

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state);

    let request = Request::builder()
        .method(Method::DELETE)
        .uri(format!("/names/{}", delete_name_id))
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    // Should show remaining names (2) and updated count
    assert!(body_text.contains("TestUser1"));
    assert!(body_text.contains("TestUser2"));
    assert!(!body_text.contains("DeleteTestUser"));
    assert!(body_text.contains("<div class=\"stat-value\">2</div>"));

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "delete_name_and_update_table_count",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_handle_delete_request_for_nonexistent_name() {
    let state = setup().await.expect("Failed to setup test context");

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state);

    // Try to delete a name with ID that doesn't exist
    let request = Request::builder()
        .method(Method::DELETE)
        .uri("/names/99999")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    // Should return an error status
    assert_eq!(response.status(), StatusCode::INTERNAL_SERVER_ERROR);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    // Should contain error message
    assert!(body_text.contains("An unexpected error occurred"));

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::INTERNAL_SERVER_ERROR,
        Some("text/html; charset=utf-8"),
        &headers,
        "delete_nonexistent_name_error",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn delete_endpoint_returns_table_fragment_not_full_page() {
    let state = setup().await.expect("Failed to setup test context");
    let name_id = create_single_test_name(&state.db).await;

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state);

    let request = Request::builder()
        .method(Method::DELETE)
        .uri(format!("/names/{}", name_id))
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    // Should return only the table fragment, not a full HTML page
    assert!(!body_text.contains("<html"));
    assert!(!body_text.contains("<head"));
    assert!(!body_text.contains("<body"));

    // Should contain the table structure or empty message
    assert!(body_text.contains("No names found") || body_text.contains("<table"));

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "delete_returns_table_fragment",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn delete_endpoint_returns_correct_content_type() {
    let state = setup().await.expect("Failed to setup test context");
    let name_id = create_single_test_name(&state.db).await;

    let name_state = NameState {
        db: Arc::new(state.db),
    };
    let app = create_name_router(name_state);

    let request = Request::builder()
        .method(Method::DELETE)
        .uri(format!("/names/{}", name_id))
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);
    assert_eq!(
        response.headers().get("content-type").unwrap(),
        "text/html; charset=utf-8"
    );

    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        StatusCode::OK,
        Some("text/html; charset=utf-8"),
        &headers,
        "delete_endpoint_content_type_check",
    );

    assert_yaml_snapshot!(snapshot_data);
}
