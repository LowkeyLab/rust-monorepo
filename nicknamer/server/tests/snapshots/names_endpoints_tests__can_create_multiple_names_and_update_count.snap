---
source: nicknamer/server/tests/names_endpoints_tests.rs
expression: snapshot_data
---
test_context: create_multiple_names_update_count
status: 200
headers:
  content-type: text/html; charset=utf-8
html_body:
  - ""
  - "<form"
  - "  id=\"bulk-delete-form\""
  - "  hx-delete=\"/names\""
  - "  hx-target=\"#names-table\""
  - "  hx-confirm=\"Are you sure you want to delete the selected names?\""
  - ">"
  - "  <div class=\"mb-4 flex justify-between items-center\">"
  - "    <button"
  - "      type=\"submit\""
  - "      class=\"btn btn-error\""
  - "      id=\"delete-selected-btn\""
  - "      disabled"
  - "    >"
  - "      <svg"
  - "        xmlns=\"http://www.w3.org/2000/svg\""
  - "        class=\"h-4 w-4 mr-2\""
  - "        fill=\"none\""
  - "        viewBox=\"0 0 24 24\""
  - "        stroke=\"currentColor\""
  - "      >"
  - "        <path"
  - "          stroke-linecap=\"round\""
  - "          stroke-linejoin=\"round\""
  - "          stroke-width=\"2\""
  - "          d=\"M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1-1v3M4 7h16\""
  - "        />"
  - "      </svg>"
  - "      Delete Selected"
  - "    </button>"
  - "  </div>"
  - ""
  - "  <div class=\"overflow-x-auto\">"
  - "    <table class=\"table table-zebra w-full table-fixed\">"
  - "      <thead>"
  - "        <tr>"
  - "          <th class=\"w-16\">"
  - "            <label>"
  - "              <input"
  - "                type=\"checkbox\""
  - "                class=\"checkbox\""
  - "                id=\"select-all\""
  - "                onchange=\"toggleAllCheckboxes(this)\""
  - "              />"
  - "            </label>"
  - "          </th>"
  - "          <th class=\"w-32\">Discord ID</th>"
  - "          <th class=\"w-48\">Name</th>"
  - "          <th class=\"w-32\">Server ID</th>"
  - "          <th class=\"w-40\"></th>"
  - "        </tr>"
  - "      </thead>"
  - "      <tbody>"
  - "         <tr id=\"name-row-1\">"
  - "  <td>"
  - "    <label>"
  - "      <input"
  - "        type=\"checkbox\""
  - "        class=\"checkbox\""
  - "        name=\"selected_ids\""
  - "        value=\"1\""
  - "        onchange=\"updateDeleteButton()\""
  - "      />"
  - "    </label>"
  - "  </td>"
  - "  <td>123456789</td>"
  - "  <td class=\"font-semibold\">TestUser1</td>"
  - "  <td>test-server-1</td>"
  - "  <th>"
  - "    <button"
  - "      class=\"btn btn-warning btn-sm mr-2\""
  - "      hx-get=\"/names/1/edit\""
  - "      hx-target=\"closest tr\""
  - "      hx-swap=\"outerHTML\""
  - "    >"
  - "      <svg"
  - "        xmlns=\"http://www.w3.org/2000/svg\""
  - "        class=\"h-4 w-4\""
  - "        fill=\"none\""
  - "        viewBox=\"0 0 24 24\""
  - "        stroke=\"currentColor\""
  - "      >"
  - "        <path"
  - "          stroke-linecap=\"round\""
  - "          stroke-linejoin=\"round\""
  - "          stroke-width=\"2\""
  - "          d=\"M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z\""
  - "        />"
  - "      </svg>"
  - "      Edit"
  - "    </button>"
  - "    <button"
  - "      class=\"btn btn-error btn-sm\""
  - "      hx-delete=\"/names/1\""
  - "      hx-target=\"#names-table\""
  - "      hx-confirm=\"Are you sure you want to delete this name entry?\""
  - "    >"
  - "      <svg"
  - "        xmlns=\"http://www.w3.org/2000/svg\""
  - "        class=\"h-4 w-4\""
  - "        fill=\"none\""
  - "        viewBox=\"0 0 24 24\""
  - "        stroke=\"currentColor\""
  - "      >"
  - "        <path"
  - "          stroke-linecap=\"round\""
  - "          stroke-linejoin=\"round\""
  - "          stroke-width=\"2\""
  - "          d=\"M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16\""
  - "        />"
  - "      </svg>"
  - "      Delete"
  - "    </button>"
  - "  </th>"
  - "</tr>  <tr id=\"name-row-2\">"
  - "  <td>"
  - "    <label>"
  - "      <input"
  - "        type=\"checkbox\""
  - "        class=\"checkbox\""
  - "        name=\"selected_ids\""
  - "        value=\"2\""
  - "        onchange=\"updateDeleteButton()\""
  - "      />"
  - "    </label>"
  - "  </td>"
  - "  <td>987654321</td>"
  - "  <td class=\"font-semibold\">TestUser2</td>"
  - "  <td>test-server-1</td>"
  - "  <th>"
  - "    <button"
  - "      class=\"btn btn-warning btn-sm mr-2\""
  - "      hx-get=\"/names/2/edit\""
  - "      hx-target=\"closest tr\""
  - "      hx-swap=\"outerHTML\""
  - "    >"
  - "      <svg"
  - "        xmlns=\"http://www.w3.org/2000/svg\""
  - "        class=\"h-4 w-4\""
  - "        fill=\"none\""
  - "        viewBox=\"0 0 24 24\""
  - "        stroke=\"currentColor\""
  - "      >"
  - "        <path"
  - "          stroke-linecap=\"round\""
  - "          stroke-linejoin=\"round\""
  - "          stroke-width=\"2\""
  - "          d=\"M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z\""
  - "        />"
  - "      </svg>"
  - "      Edit"
  - "    </button>"
  - "    <button"
  - "      class=\"btn btn-error btn-sm\""
  - "      hx-delete=\"/names/2\""
  - "      hx-target=\"#names-table\""
  - "      hx-confirm=\"Are you sure you want to delete this name entry?\""
  - "    >"
  - "      <svg"
  - "        xmlns=\"http://www.w3.org/2000/svg\""
  - "        class=\"h-4 w-4\""
  - "        fill=\"none\""
  - "        viewBox=\"0 0 24 24\""
  - "        stroke=\"currentColor\""
  - "      >"
  - "        <path"
  - "          stroke-linecap=\"round\""
  - "          stroke-linejoin=\"round\""
  - "          stroke-width=\"2\""
  - "          d=\"M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16\""
  - "        />"
  - "      </svg>"
  - "      Delete"
  - "    </button>"
  - "  </th>"
  - "</tr>  <tr id=\"name-row-3\">"
  - "  <td>"
  - "    <label>"
  - "      <input"
  - "        type=\"checkbox\""
  - "        class=\"checkbox\""
  - "        name=\"selected_ids\""
  - "        value=\"3\""
  - "        onchange=\"updateDeleteButton()\""
  - "      />"
  - "    </label>"
  - "  </td>"
  - "  <td>111222333</td>"
  - "  <td class=\"font-semibold\">ThirdUser</td>"
  - "  <td>test-server-1</td>"
  - "  <th>"
  - "    <button"
  - "      class=\"btn btn-warning btn-sm mr-2\""
  - "      hx-get=\"/names/3/edit\""
  - "      hx-target=\"closest tr\""
  - "      hx-swap=\"outerHTML\""
  - "    >"
  - "      <svg"
  - "        xmlns=\"http://www.w3.org/2000/svg\""
  - "        class=\"h-4 w-4\""
  - "        fill=\"none\""
  - "        viewBox=\"0 0 24 24\""
  - "        stroke=\"currentColor\""
  - "      >"
  - "        <path"
  - "          stroke-linecap=\"round\""
  - "          stroke-linejoin=\"round\""
  - "          stroke-width=\"2\""
  - "          d=\"M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z\""
  - "        />"
  - "      </svg>"
  - "      Edit"
  - "    </button>"
  - "    <button"
  - "      class=\"btn btn-error btn-sm\""
  - "      hx-delete=\"/names/3\""
  - "      hx-target=\"#names-table\""
  - "      hx-confirm=\"Are you sure you want to delete this name entry?\""
  - "    >"
  - "      <svg"
  - "        xmlns=\"http://www.w3.org/2000/svg\""
  - "        class=\"h-4 w-4\""
  - "        fill=\"none\""
  - "        viewBox=\"0 0 24 24\""
  - "        stroke=\"currentColor\""
  - "      >"
  - "        <path"
  - "          stroke-linecap=\"round\""
  - "          stroke-linejoin=\"round\""
  - "          stroke-width=\"2\""
  - "          d=\"M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16\""
  - "        />"
  - "      </svg>"
  - "      Delete"
  - "    </button>"
  - "  </th>"
  - "</tr> "
  - "      </tbody>"
  - "    </table>"
  - "  </div>"
  - "</form>"
  - ""
  - "<script>"
  - "  function toggleAllCheckboxes(selectAllCheckbox) {"
  - "    const checkboxes = document.querySelectorAll('input[name=\"selected_ids\"]');"
  - "    checkboxes.forEach((checkbox) => {"
  - "      checkbox.checked = selectAllCheckbox.checked;"
  - "    });"
  - "    updateDeleteButton();"
  - "  }"
  - ""
  - "  function updateDeleteButton() {"
  - "    const checkboxes = document.querySelectorAll("
  - "      'input[name=\"selected_ids\"]:checked'"
  - "    );"
  - "    const deleteBtn = document.getElementById(\"delete-selected-btn\");"
  - "    deleteBtn.disabled = checkboxes.length === 0;"
  - "  }"
  - ""
  - "  // Add event listeners to individual checkboxes"
  - "  document.addEventListener(\"DOMContentLoaded\", function () {"
  - "    const checkboxes = document.querySelectorAll('input[name=\"selected_ids\"]');"
  - "    checkboxes.forEach((checkbox) => {"
  - "      checkbox.addEventListener(\"change\", updateDeleteButton);"
  - "    });"
  - "    updateDeleteButton();"
  - "  });"
  - "</script>"
  - ""
  - ""
  - ""
  - "<div class=\"stats stats-horizontal shadow mt-6\">"
  - "  <div class=\"stat\">"
  - "    <div class=\"stat-title\">Total Names</div>"
  - "    <div class=\"stat-value\">3</div>"
  - "    <div class=\"stat-desc\">Names in database</div>"
  - "  </div>"
  - "</div>"
